packages/shopify-app-remix/src/server/authenticate/admin/embedded-authenticate.ts

```diff
@@ -0,0 +1,245 @@
+      url.pathname === config.auth.patchSessionTokenPath;
+    const isExitIframe = url.pathname === config.auth.exitIframePath;
+    const isAuthRequest = url.pathname === config.auth.path;
+
+    if (isPatchSessionToken) {
+      logger.debug('Rendering bounce page');
+      throw renderAppBridge(params, request);
+    } else if (isExitIframe) {
+      const destination = url.searchParams.get('exitIframe')!;
+
+      logger.debug('Rendering exit iframe page', {destination});
+      throw renderAppBridge(params, request, {url: destination});
+    } else if (isAuthRequest) {
+      const shop = ensureValidShopParam(request, {api, logger, config});
+      const installUrl = `https://${shop}/admin/oauth/install?client_id=${config.apiKey}`;
+      throw redirect(installUrl);
+    }
+
+    const sessionTokenHeader = getSessionTokenHeader(request);
+    const sessionTokenParam = url.searchParams.get(SESSION_TOKEN_PARAM)!;
+    const sessionTokenString = sessionTokenHeader || sessionTokenParam;
+
+    logger.info('Authenticating admin request');
+
+    if (sessionTokenParam) {
+      await validateUrlParams(request, {api, logger, config});
+      await ensureAppIsEmbeddedIfRequired(request, {api, logger, config});
+    }
+
+    if (sessionTokenString) {
+      try {
+        const sessionToken = await validateSessionTokenUncaught(
+          {api, logger, config},
+          sessionTokenString,
+        );
+
+        return await this.getAccessToken(
+          request,
+          sessionTokenString,
+          sessionToken,
+        );
+      } catch (error) {
+        if (
+          error instanceof InvalidJwtError ||
+          (error instanceof HttpResponseError &&
+            error.response.code === 400 &&
+            error.response.body?.error === 'invalid_subject_token')
+        ) {
+          console.log(`CAUGHT ${JSON.stringify(error)}`);
+          throw this.handleInvalidSession(request);
+        }
+        throw error;
+      }
+    } else {
+      logger.debug(
+        'Missing session token in search params, going to bounce page',
+      );
+      throw this.handleInvalidSession(request);
+    }
+  }
+
+  private async getAccessToken(
+    request: Request,
+    sessionToken: string,
+    payload: JwtPayload,
+  ) {
+    const {config, logger, api} = this;
+
+    const sessionId = await api.session.getCurrentId({
+      isOnline: config.useOnlineTokens,
+      rawRequest: request,
+    });
+
+    const dest = new URL(payload.dest);
+    const shop = dest.hostname;
+
+    if (sessionId) {
+      logger.debug(`SESSION ID: ${sessionId}`);
+      const persistedSession = await config.sessionStorage.loadSession(
+        sessionId,
+      );
+
+      if (persistedSession) {
+        logger.debug(`Reusing existing token: ${persistedSession.accessToken}`);
+
+        if (persistedSession.isScopeChanged(config.scopes)) {
+          config.sessionStorage.deleteSession(persistedSession.id);
+
+          this.redirectToInstall(request, shop);
```

Comment:
This is an edge case that can happen when the app changes scopes, while a merchant has an active app browser session. Please see the [discussion for more details](https://github.com/Shopify/develop-app-access/discussions/92).